Made petition campaign_id a required field

danielfmiranda opened the pull request:
# Description

<!-- Describe the PR here -->
Related PRs/issues: #10579


**This PR does the following:**
- Makes the petition snippet's `campaign_id` field required. If a user tries to create or edit a petition without filling out this field, they will recieve a validation error.
- Updates `petition-thank-you.jsx` to pass the newly required `campaign_id` as a prop. The campaign ID is then appended to the donate modal's "donate_url". (Ex: `?form=donate&c_id=<id>`). That way FRU can use this ID for reporting purposes.
- Updates `002-thank-you.spec.js` to account for newly appended URL parameter.


# Screenshots:

**URL when hovering over "Yes, Ill chip in" Button:**

<img width="951" alt="Screenshot 2023-07-18 at 5 34 04 PM" src="https://github.com/MozillaFoundation/foundation.mozilla.org/assets/18314510/4cf175e8-65ef-491b-9e22-1dee8a1412cf">

**GTM FRU Donate Modal opened event:**

![image](https://github.com/MozillaFoundation/foundation.mozilla.org/assets/18314510/a7d01319-ce62-49fc-b79d-05dc516fff51)



# Steps to test

## Testing that campaign_id is required

1. Copy the prod db on your local machine
2. Visit http://localhost:8000/cms/snippets/wagtailpages/petition/edit/306/
3. Delete what is set for `campaign_id` and click "Publish"
4. You should receive a validation error stating that Campaign ID is required


## Testing that the campaign_id is being sent to FRU/Salesforce
1. Check this branch out on your local machine
5. run `inv copy-prod-db`
6. Log into [google tag manager](https://tagmanager.google.com/#/container/accounts/6002697367/containers/36502666/workspaces/22) and click "preview"
7. In the URL box, enter http://localhost:8000/en/campaigns/tell-slack-make-dms-safe/?thank_you=true
8. On the new tab that appears, hover over the "Yes Ill chip in" button, note the `c_id` that is shown in the URL preview at the bottom left of the window.
9. Click the "Yes, I'll chip in" button to render the modal.
10. Go back to the GTM screen, and click the event " Event: FundraiseUp.checkoutOpen" on the left hand side of the screen, under "Summary".
11. Expand the details, the c_id should match the c_id that was shown on step five.


<img width="400" alt="Screenshot 2023-07-19 at 11 04 00 AM" src="https://github.com/MozillaFoundation/foundation.mozilla.org/assets/18314510/e62fc733-e72d-4387-a20c-9e1f28554f7a">


9. If everything is working as expected, testing is complete!

danielfmiranda commented:
Hi @mtdenton! 

In this PR I am appending the campaign ID parameter as `campaign_id` however after looking further into [FRU documentation ](https://fundraiseup.com/support/utm-parameters/)and [Salesforce Documentation](https://help.salesforce.com/s/articleView?id=sf.mc_gai_utm_parameters.htm&type=5), it appears that they are both referencing the parameter as `utm_campaign`. 

I was wondering if you knew who I could reach out to to confirm what the param should be appended as? Thanks!

danielfmiranda commented on source/js/components/petition/donation-modal.jsx:
```diff
@@ -85,6 +86,8 @@ class DonationModal extends Component {
     if (!this.donateURL) {
```
There seems to be no reference in the code base regarding donateURL as a prop. So at the moment this link is always defaulting to `form=donate`. However, I am leaving this here as I can use this prop while implementing #10846

mmmavis replied:
Thanks for the note. Either in this PR or in #10846, could we add a method for generating `donateURL`? Right now the variable name `donateURL` could be a bit misleading as it's default to a query string with `?` and later appended with more query params. I think using [`URLSearchParams`](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams/URLSearchParams) would probably be a good/cleaner choice.

mtdenton commented:
Hey Daniel! The UTM campaign and the Salesforce campaign are two separate things to track and send along here. The expectation is that you would have a Salesforce campaign ID to send through to the FRU modal 100% of the time. You would also send `utm_campaign` along with any other accompanying `utm_xx` parameters that come through on the page's URL if they're present.

Just looking at the above screenshot as well, I'm seeing `c_id` and `campaign_id` and they both have some sort of Salesforce-style ID. Where is `c_id` coming from?

danielfmiranda commented:
Great point @mtdenton I did not catch that! 

After taking a look inside FRU, it looks like that is a parameter that is set through FRU itself for our campaign titled "Mofo Main":

![Screenshot 2023-07-19 at 10-57-02 Fundraise Up](https://github.com/MozillaFoundation/foundation.mozilla.org/assets/18314510/df442d6b-965c-4281-a477-49971f917c2c)

danielfmiranda commented on tests/integration/petition/002-thank-you.spec.js:
```diff
@@ -44,7 +44,7 @@ test.describe("Donation modal", () => {
     // test if FRU iframe pops up after clicking the Yes button
     const yesDonateButton = page.locator(
-      `.modal-content .tw-btn-primary[href="?form=donate"]`
+      `.modal-content .tw-btn-primary[href*="?form=donate&c_id="]`
```
Making sure that the donate URL includes the `c_id` URL param.

mtdenton commented:
Aha, good find @danielfmiranda. Let's get some stakeholder input here to make sure that's the intent. I'll have an answer soon

danielfmiranda commented:
Hi @mtdenton

I have confirmed with Esteban that parameter `c_id` is in fact what we are looking for. Now submitting this PR for review!

mmmavis commented on network-api/networkapi/wagtailpages/pagemodels/campaigns.py:
```diff
@@ -171,7 +171,6 @@ class Petition(TranslatableMixin, CTA):
         max_length=20,
         help_text="Which campaign identifier should this petition be tied to?",
```
Since we are updating this field, could we also update the help text to have "Salesforce" for extra clarity? e.g., something like `"Which Salesforce campaign identifier should this petition be tied to?"`
```suggestion
        help_text="Which Salesforce campaign identifier should this petition be tied to?",
```

danielfmiranda replied:
Thanks @mmmavis! This is a great suggestion and has been implemented now.

mmmavis reviewed (changes requested):
Thanks for the inline comments, @danielfmiranda .It makes PR review a lot easier! I left some comments but they are mostly for adding extra clarity. We'll also need to update `PETITION_TEST_CAMPAIGN_ID` to have a non-empty default value since we are changing the field to be required now :)

danielfmiranda commented:
Thanks for the feedback @mmmavis! I have updated the help text to include "Salesforce" in the sentence now, and have also updated the env.default file to include the sandbox Salesforce campaign ID that was discussed in Slack.

Regarding the feedback about the `donateURL` I would vote to skip it in this PR, and I will take care of it in #10846 as I am going to be making changes to the donate URL in that ticket.

I have now requested review, thanks again!

mmmavis commented on env.default:
```diff
@@ -62,7 +62,7 @@ CSP_INCLUDE_NONCE_IN=script-src
 
 
 # TEST ENVIRONMENT VALUES
```
```suggestion
# Petition test campaign id for Salesforce Sandbox
```

mmmavis commented on network-api/networkapi/wagtailpages/migrations/0094_make_petition_id_required.py:
```diff
@@ -0,0 +1,22 @@
+            model_name="petition",
+            name="campaign_id",
+            field=models.CharField(
+                help_text="Which Salesforce campaign identifier should this petition be tied to?",
```
```suggestion
                help_text="Which Salesforce Campaign ID should this petition be tied to?",
```

Per new request on dev Slack channel

mmmavis commented on network-api/networkapi/wagtailpages/pagemodels/campaigns.py:
```diff
@@ -169,9 +169,8 @@ class Meta:
     campaign_id = models.CharField(
         max_length=20,
-        help_text="Which campaign identifier should this petition be tied to?",
+        help_text="Which Salesforce campaign identifier should this petition be tied to?",
```
```suggestion
        help_text="Which Salesforce Campaign ID should this petition be tied to?",
```

mmmavis reviewed (approved):
PR looks good! There are a few quick changes we can make to add more clarity around petition flow. I've left code suggestions inline. Feel free to modify them or just commit my suggestions. Either way we should squash commits for merging so it doesn't create issues with migration file.

Thanks @danielfmiranda !
